File: jApp/src/main/java/j/s/yarlykov/ui/fragmentbased/CitiesFragment.java
/**
 * Materials:
 * http://developer.alexanderklimov.ru/android/listfragment.php
 *
 *  { "id": 498817, "name": "Saint Petersburg", "country": "RU", "coord": { "lon": 30.264168, "lat": 59.894444 }
 *
 */

package j.s.yarlykov.ui.fragmentbased;

import android.content.res.Configuration;
import android.content.res.TypedArray;
import android.os.Bundle;
import android.os.IBinder;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.app.ListFragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ListView;
import android.widget.SimpleAdapter;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import j.s.yarlykov.R;

public class CitiesFragment extends ListFragment {

    private boolean isLandscape;
    private int selectedPosition = 0;
    private final String selectedPositionKey = "SelectedCity";
    private ListView listView;

    private final String KEY_IMAGE_ID = "image";
    private final String KEY_CITY = "city";
    private final static String KEY_BINDER = "binder";
    private IBinder forecastSource = null;

    public static CitiesFragment create(IBinder binder) {
        CitiesFragment fragment = new CitiesFragment();
        Bundle args = new Bundle();
        args.putBinder(KEY_BINDER, binder);
        fragment.setArguments(args);

        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        forecastSource = getArguments().getBinder(KEY_BINDER);
    }

    @Nullable
    @Override
    public View onCreateView(
            @NonNull LayoutInflater inflater,
            @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.cities_list_fragment, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        listView = getListView();
        initList();
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        isLandscape = getResources().getConfiguration().orientation
                == Configuration.ORIENTATION_LANDSCAPE;

        if(savedInstanceState != null){
            selectedPosition = savedInstanceState.getInt(selectedPositionKey, 0);
        }

        if(isLandscape){
            listView.setChoiceMode(ListView.CHOICE_MODE_SINGLE);
            listView.setItemChecked(selectedPosition, true);
            showForecast();
        }
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        outState.putInt(selectedPositionKey, selectedPosition);
        super.onSaveInstanceState(outState);
    }

    private void initList() {
        ArrayList<Map<String, Object>> data = new ArrayList<>();
        TypedArray images = getResources().obtainTypedArray(R.array.armsImages);
        String[] cities = getResources().getStringArray(R.array.cities);

        Map<String, Object> map;

        for(int i = 0; i < cities.length; i++) {
            map = new HashMap<>();
            map.put(KEY_IMAGE_ID, images.getResourceId(i, 0));
            map.put(KEY_CITY, cities[i]);
            data.add(map);
        }
        images.recycle();

        String[] from = {KEY_IMAGE_ID, KEY_CITY};
        int[] to = {R.id.ivArms, R.id.tvCity};

        SimpleAdapter sAdapter = new SimpleAdapter(
                getActivity(),
                data,
                R.layout.cities_list_item,
                from,
                to);

        setListAdapter(sAdapter);

        listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                selectedPosition = position;
                showForecast();
            }
        });
    }

    private void showForecast(){
        String citySelected = getResources().getStringArray(R.array.cities_openweather)[selectedPosition];

        if(isLandscape) {
            getActivity().findViewById(R.id.rightFrame).setVisibility(View.VISIBLE);
            listView.setItemChecked(selectedPosition, true);

            ForecastFragment forecastFragment = (ForecastFragment)getFragmentManager()
                    .findFragmentById(R.id.rightFrame);

            if(forecastFragment == null || forecastFragment.getIndex() != selectedPosition) {

                forecastFragment = ForecastFragment.create(forecastSource, citySelected, selectedPosition);

                FragmentTransaction ft = getFragmentManager().beginTransaction();
                ft.replace(R.id.rightFrame, forecastFragment);
                ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE);
                ft.commit();
            }
        } else {
            ForecastActivityFr.start(getContext(), forecastSource, citySelected, selectedPosition);
        }
    }
}

File: kApp/src/main/java/k/s/yarlykov/ui/fragmentbased/CitiesFragment.kt
package k.s.yarlykov.ui.fragmentbased

import android.content.res.Configuration
import android.os.Binder
import android.os.Bundle
import android.os.IBinder
import android.support.v4.app.FragmentTransaction
import android.support.v4.app.ListFragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.FrameLayout
import android.widget.ListView
import android.widget.SimpleAdapter
import k.s.yarlykov.R
import k.s.yarlykov.data.domain.CityForecast
import k.s.yarlykov.data.domain.Forecast
import k.s.yarlykov.services.ForecastService
import java.util.*

class CitiesFragment : ListFragment() {

    companion object {
        val KEY_BINDER = "binder"

        fun create(binder: IBinder): CitiesFragment {
            return CitiesFragment().apply {
                arguments = Bundle().apply {
                    putBinder(KEY_BINDER, binder)
                }
            }
        }
    }

    interface ForecastSource {
        fun getForecastById(id: Int): Forecast
        fun getForecastByCity(city: String): Forecast?
    }

    private  val selectedPositionKey = "SelectedCity"
    private  val KEY_IMAGE_ID = "image"
    private  val KEY_CITY = "city"

    private var isLandscape: Boolean = false
    private var selectedPosition: Int = 0
    private var forecastSource: IBinder? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        forecastSource =  arguments?.getBinder(KEY_BINDER)
    }

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.cities_list_fragment, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initList()
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        isLandscape = resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE

        savedInstanceState?.let {
            selectedPosition = it.getInt(selectedPositionKey, 0)
        }

        if (isLandscape) {
            listView.choiceMode = ListView.CHOICE_MODE_SINGLE
            listView.setItemChecked(selectedPosition, true)
            showForecast()
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putInt(selectedPositionKey, selectedPosition)
        super.onSaveInstanceState(outState)
    }

    private fun initList() {
        val data = ArrayList<Map<String, Any>>()
        val images = resources.obtainTypedArray(R.array.armsImages)
        val cities = resources.getStringArray(R.array.cities)

        for ((i, city) in cities.withIndex()) {
            val map = mapOf(
                    KEY_IMAGE_ID to images.getResourceId(i, 0),
                    KEY_CITY to city)
            data.add(map)
        }
        images.recycle()

        val from = arrayOf(KEY_IMAGE_ID, KEY_CITY)
        val to = intArrayOf(R.id.ivArms, R.id.tvCity)

        listAdapter = SimpleAdapter(
                activity,
                data,
                R.layout.cities_list_item,
                from,
                to)

        listView.onItemClickListener = AdapterView.OnItemClickListener { _, _, position, _ ->
            selectedPosition = position
            showForecast()
        }
    }

    private fun showForecast() {
        val citySelected = resources.getStringArray(R.array.cities_openweather)[selectedPosition]

        if (isLandscape) {

            activity!!.findViewById<FrameLayout>(R.id.rightFrame).visibility = View.VISIBLE
            listView.setItemChecked(selectedPosition, true)

            var forecastFragment = fragmentManager!!
                    .findFragmentById(R.id.rightFrame) as ForecastFragment?

            if (forecastFragment == null || forecastFragment.getIndex() != selectedPosition) {
//                forecastFragment = ForecastFragment.create(forecastSource, citySelected, selectedPosition)

                val ft = fragmentManager!!.beginTransaction()
//                ft.replace(R.id.rightFrame, forecastFragment)
                ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                ft.commit()
            }
        } else {
            ForecastActivityFr.start(context, forecastSource, citySelected, selectedPosition)
        }
    }
}